Stabilize or_patterns (RFC 2535, 2530, 2175)

mark-i-m opened the pull request:
closes #54883

This PR stabilizes the or_patterns feature in Rust 1.53.

This is blocked on the following (in order):
- [x] The crater run in https://github.com/rust-lang/rust/pull/78935#issuecomment-731564021
- [x] The resolution of the unresolved questions and a second crater run (https://github.com/rust-lang/rust/pull/78935#issuecomment-735412705)
    - It looks like we will need to pursue some sort of edition-based transition for `:pat`.
- [x] Nomination and discussion by T-lang
- [x] Implement new behavior for `:pat` based on consensus (https://github.com/rust-lang/rust/pull/80100).
- [ ] An FCP on stabilization

EDIT: Stabilization report is in https://github.com/rust-lang/rust/pull/79278#issuecomment-772815177

rust-highfive commented:
r? @estebank

(rust_highfive has picked a reviewer for you, use r? to override)

mark-i-m commented:
r? @petrochenkov

mark-i-m commented:
I'm not sure I understand the CI failure... I thought `cfg_attr(bootstrap...)` wasn't needed in the compiler?

est31 commented:
@mark-i-m it's only not needed when you stabilize a std feature as std is always compiled first.

mark-i-m commented:
@est31 Do you happen to know how to bless clippy tests? Everything is passing locally on stage 2.

est31 commented:
@mark-i-m sorry no idea off the top of my head. IIRC the compiletest code creates the files on the disk and you only have to copy them to the right location. bless is only a wrapper around that, but no idea whether clippy has better facilities than manual copies.

camelid commented:
> `@est31` Do you happen to know how to bless clippy tests? Everything is passing locally on stage 2.

Unfortunately you can't bless Clippy tests in-tree :/

I just [asked about this the other day in #clippy](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/blessing.20clippy.20tests.20in-tree).

mark-i-m commented:
@camelid I see. So what should one do with the CI failure? Does it just stay that way?

camelid commented:
I think you might need to update by hand :/

bors commented:
:umbrella: The latest upstream changes (presumably #79336) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

bors commented:
:umbrella: The latest upstream changes (presumably #79364) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

bors commented:
:umbrella: The latest upstream changes (presumably #79927) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1583961283) [(plain)](https://github.com/rust-lang/rust/commit/dbb5f6b128f4418e615e9cfe0cdfd8d93a8a44b7/checks/1583961283/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9000/11186
.................................................................................................... 9100/11186
.............................................................................i......i............... 9200/11186
.................................................................................................... 9300/11186
................iiiiii..iiiiii.i.................................................................... 9400/11186
.................................................................................................... 9600/11186
.................................................................................................... 9700/11186
.................................................................................................... 9800/11186
.................................................................................................... 9900/11186
---
97 error: unreachable pattern
-   --> $DIR/exhaustiveness-unreachable-pattern.rs:68:14
+   --> $DIR/exhaustiveness-unreachable-pattern.rs:74:10
99    |
+ LL |         [true
+ 
+ error: unreachable pattern
+   --> $DIR/exhaustiveness-unreachable-pattern.rs:81:36
+    |
+    |
+ LL |         (true | false, None | Some(true
+ 
+ error: unreachable pattern
+   --> $DIR/exhaustiveness-unreachable-pattern.rs:97:14
+    |
---
103 error: unreachable pattern
-   --> $DIR/exhaustiveness-unreachable-pattern.rs:87:19
+   --> $DIR/exhaustiveness-unreachable-pattern.rs:116:19
105    |
106 LL |                 | false) => {}

108 
109 error: unreachable pattern
-   --> $DIR/exhaustiveness-unreachable-pattern.rs:95:15
---
119    |               ^^^^


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/or-patterns/exhaustiveness-unreachable-pattern/exhaustiveness-unreachable-pattern.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args or-patterns/exhaustiveness-unreachable-pattern.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/or-patterns/exhaustiveness-unreachable-pattern" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/or-patterns/exhaustiveness-unreachable-pattern/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:7:9
   |
LL |         (1,) => {} //~ ERROR unreachable pattern
   |
note: the lint level is defined here
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:1:9
   |
   |
LL | #![deny(unreachable_patterns)]

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:12:9
   |
   |
LL |         (2,) => {} //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:18:9
   |
   |
LL |         (1 | 2,) => {} //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:23:9
   |
   |
LL |         (1, 3) => {}     //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:24:9
   |
   |
LL |         (1, 4) => {}     //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:25:9
   |
   |
LL |         (2, 4) => {}     //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:26:9
   |
   |
LL |         (2 | 1, 4) => {} //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:28:9
   |
   |
LL |         (1, 4 | 5) => {} //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:36:9
   |
   |
LL |         (Some(1),) => {} //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:37:9
   |
   |
LL |         (None,) => {}    //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:42:9
   |
   |
LL |         ((1..=4,),) => {} //~ ERROR unreachable pattern

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:47:14
   |
   |
LL |         (1 | 1,) => {} //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:52:15
   |
   |
LL |             | 0 //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:54:15
   |
   |
LL |             | 0] => {} //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:62:10
   |
   |
LL |         [1 //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:74:10
   |
   |
LL |         [true //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:81:36
   |
   |
LL |         (true | false, None | Some(true //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:97:14
   |
   |
LL |         Some(0 //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:116:19
   |
   |
LL |                 | false) => {} //~ ERROR unreachable

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:124:15
   |
   |
LL |             | true) => {} //~ ERROR unreachable
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu

error: unreachable pattern
  --> /checkout/src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:130:15
   |
   |
LL |             | true, //~ ERROR unreachable

error: aborting due to 21 previous errors


---
test result: FAILED. 11101 passed; 1 failed; 84 ignored; 0 measured; 0 filtered out; finished in 146.56s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:16:22
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #79135) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

mark-i-m commented:
AFAICT this is waiting on hearing back from a team member and maybe having an FCP.

@rustbot modify labels: -S-blocked +S-waiting-on-review

bors commented:
:umbrella: The latest upstream changes (presumably #80511) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1631640338) [(plain)](https://github.com/rust-lang/rust/commit/5a3e8034e593d4f930c53f64a2e7d64bb20e2b2f/checks/1631640338/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9000/11228
.................................................................................................... 9100/11228
.................................................................................................... 9200/11228
...............i......i............................................................................. 9300/11228
......................................................iiiiii..iiiiii.i.............................. 9400/11228
.................................................................................................... 9600/11228
.................................................................................................... 9700/11228
.................................................................................................... 9800/11228
.................................................................................................... 9900/11228
---
failures:

---- [ui] ui/macros/edition-macro-pats.rs stdout ----
normalized stderr:
warning: the feature `or_patterns` has been stable since 1.51.0 and no longer requires an attribute to enable
   |
LL | #![feature(or_patterns)]
   |            ^^^^^^^^^^^
   |
---



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/edition-macro-pats/edition-macro-pats.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args macros/edition-macro-pats.rs`
error: 1 errors occurred comparing output.
status: exit code: 0
status: exit code: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/macros/edition-macro-pats.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/edition-macro-pats/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/edition-macro-pats/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: the feature `or_patterns` has been stable since 1.51.0 and no longer requires an attribute to enable
   |
LL | #![feature(or_patterns)]
   |            ^^^^^^^^^^^
   |
---

Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:16:18
```

</details>

joshtriplett commented:
We discussed this in today's @rust-lang/lang meeting, and we're prepared to start FCP to merge/stabilize as soon as we have a stabilization report.

mark-i-m commented:
Ah, yeah, I forgot about that. Unfortunately, I have neither the knowledge nor the time to produce the full report atm. I would need help from someone else who knows more  (perhaps @varkor or @Nadrieril ?). 


> The report should contain:
>
>    A summary, showing examples (e.g. code snippets) what is enabled by this feature.
>    Links to test cases in our test suite regarding this feature and describe the feature's behavior on encountering edge cases.
>    Links to the documentations (the PRs we have made in the previous steps).
>    Any other relevant information(Examples of such reports can be found in rust-lang/rust#44494 and rust-lang/rust#28237).
>    The resolutions of any unresolved questions if the stabilization is for an RFC.

I can write about the first and last points, but don't know much about the others...

Should I start a hackmd or something?

mark-i-m commented:
Started this hackmd: https://hackmd.io/oczPRxyBSKOAojC1Dgxupw

nikomatsakis commented:
One other note regarding stabilization:

Can we avoid stabilize the pat2018/pat2021 matchers until the edition is stabilized, just in case we slip in the edition and want to update the names?

nikomatsakis commented:
Also, once this is stabilized, we need to update the grammar in the reference. I'm not sure what protocol we usually follow... I think maybe the best practice is to have a pending PR that gets merged after we accept this one?

(cc @rust-lang/docs)

GuillaumeGomez commented:
I think this is the best course indeed. Or even open a PR referencing the one before this one is merged. I don't think the order matters much here.

bors commented:
:umbrella: The latest upstream changes (presumably #81177) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

joshtriplett commented:
> Can we avoid stabilize the pat2018/pat2021 matchers until the edition is stabilized, just in case we slip in the edition and want to update the names?

I don't think it's unreasonable to commit to calling it "Rust 2021", at this point, even if it were to get released in (for instance) early 2022.

mark-i-m commented:
> Can we avoid stabilize the pat2018/pat2021 matchers until the edition is stabilized, just in case we slip in the edition and want to update the names?

Yes, this PR does _not_ intend to stabilize these.

Sorry for the delay... I'm a bit busy for the next week or so. Then, hopefully I can come back to this.

steveklabnik commented:
>  I'm not sure what protocol we usually follow... I think maybe the best practice is to have a pending PR that gets merged after we accept this one?

We relaxed this, and the process currently is "file an issue on the reference". PRs fail until the functionality is stable, and the whole "send docs PR, merge impl PR, restart tests for docs PR, merge" dance was too much trouble.

mark-i-m commented:
I have rebased and submitted https://github.com/rust-lang/reference/pull/957 to update the reference. Since beta is branching soon, I've bumped the version in the PR to 1.52...

Also thanks to @Nadrieril for help with this report:

# Stabilization Report

This is a stabilization report for the `or_patterns` feature. It does _NOT_ stabilize the `edition_macro_pat` feature; i.e. `:pat2018/21` remain unstable.

A mighty Thank You to the many people who designed, implemented, and reviewed this feature, including (in no particular order, and apologies to any I have missed) Centril, varkor, Nadrieril, matthewjasper, petrochenkov, alexreg, estebank, davidtwco, nikomatsakis, arielb1, nnethercote, pnkfelix, nagisa, eddyb, thekuom, sapir, Nemo157, and joshtriplett :tada:

# Examples

This stabilization allows nested or-patterns in stable Rust in all of the following cases:

1. `if let` expressions:

   ```rust
   if let Foo::Bar | Foo::Quux(1 | 2) = some_computation() {
       ...
   }
   ```

1. `while let` expressions:

   ```rust
   while let Ok(1 | 2) | Err(3) = different_computation() {
       ...
   }
   ```

3. `let` statements:

   ```rust
   let Ok(x) | Err(x) = another_computation();
   ```

   In this case, the pattern must be irrefutable as `Ok(x) | Err(x)` is.

4. `fn` arguments:

   ```rust
   fn foo((Ok(x) | Err(x)): Result<u8, u8>) {
       ...
   }
   ```

   Here too, the pattern must be irrefutable.

5. `match` patterns:

   ```rust
   match foo {
       Foo::Bar | Foo::Quux(1 | 2) => { ... }
       ...
   }
   ```

It does _not_ stabilize or-patterns in closure arguments or macro `:pat` matchers (more below).

# Test cases

Test cases can be found in [the dedicated folder](https://github.com/rust-lang/rust/tree/a4cbb44ae2c80545db957763b502dc7f6ea22085/src/test/ui/or-patterns).

Additionally, the following tests test the behavior of `:pat` in different editions:
- [src/test/ui/macros/macro-pat-follow.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/macro-pat-follow.rs)
- [src/test/ui/macros/macro-pat-follow-2018.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/macro-pat-follow-2018.rs)
- [src/test/ui/macros/edition-macro-pats.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/edition-macro-pats.rs)

And the feature gate test for `edition_macro_pat`:
- [src/test/ui/feature-gate-edition_macro_pats.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/feature-gate-edition_macro_pats.rs)

# Documentation

The RFC contains the precise [grammar changes][grammar].

PR to update the reference: https://github.com/rust-lang/reference/pull/957

# Other Relevant Info

petrochekov [raised a concern](https://github.com/rust-lang/rust/issues/54883#issuecomment-742802350) about simplifying the grammar for patterns to allow a leading vert (`|`) at nested levels, not just at the top level. There was [an FCP](https://github.com/rust-lang/rust/issues/81415), and there is team consensus for doing so. However, we do not implement it in this PR because it is a backwards-compatible change. It allows strictly more code to compile. The only place where such a grammar change could be visible is in macros in 2021+, which are still unstable.

# Resolved Unresolved Questions

> Should we allow top_pat or pat<allow_top_alt> in inferrable_param such that closures permit |Ok(x) | Err(x)| without first wrapping in parenthesis?

In the [2020-11-10 T-lang meeting][tlangcon], there was consensus this should not be allowed. Despite the consistency with other places where patterns are allowed, this syntax was deemed to be too confusing.

> Should the pat macro fragment specifier match top_pat in different Rust editions or should it match pat<no_top_alt> as currently specified? We defer such decisions to stabilization because it depends on the outcome of crater runs to see what the extent of the breakage would be.

Originally, it was thought that `:pat` could match `top_pat` in all editions, but two crater runs ([0][crater0], [1][crater1]) showed that even with a fallback plan, there was lots of complex breakage.

Instead, in the [2020-12-15 T-lang meeting][tlangcon2], there was consensus that in Rust 2015/18, `:pat` would continue to match `pat<no_top_alt>` (as it already does), while in Rust 2021 and onward, it would match `top_pat`. This means that in Rust 2015/18, `:pat` will _not_ match top-level or-patterns, but in subsequent editions it will.

Additionally, to allow cross-edition macro usage, the `:pat2018` and `:pat2021` matchers were introduced behind the `edition_macro_pat` feature, which is _not_ being stabilized in this stabilization report.


[grammar]: https://github.com/rust-lang/rfcs/blob/master/text/2535-or-patterns.md#reference-level-explanation
[tlangcon]: https://github.com/rust-lang/rust/issues/54883#issuecomment-724898641
[crater0]: https://github.com/rust-lang/rust/pull/78935#issuecomment-735050394
[crater1]: https://github.com/rust-lang/rust/pull/78935#issuecomment-740828151
[tlangcon2]: https://github.com/rust-lang/rust/issues/54883#issuecomment-745509090

joshtriplett commented:
@rfcbot merge

scottmcm commented:
Hmm, I explored a bit and I'd like to bring up a bit around mixing these patterns with type annotations.

The following errors, saying that it needs parens around the pattern:
```rust
fn bar(Ok(x) | Err(x): Result<u8, u8>) {}
```

However the equivalent in a `let` doesn't warn:
```rust
let Ok(x) | Err(x): Result<u8, u8> = y;
```

1. Why are the parens needed in the first example?  I understand them as being desirable for closures (because of `|` being used in both syntaxes), but didn't see a mention of normal functions.  (Of course it's completely possible I just missed it.)
2. The RFC describes `|` in patterns as having "low precedence".  Is there any guarding that would be worth doing here in case of adding more ascription to patterns?  For example, we could allow `let Ok(x) | Err(x) = y;` but require parenthesizing with a type annotation like `let (Ok(x) | Err(x)): Result<u8, u8> = y;`.  My vague expectation would be that, if the annotations were part of the pattern grammar instead of the `let` grammar, `Ok(x) | Err(x): Result<u8, u8>` would parse as `Ok(x) | (Err(x): Result<u8, u8>)`, and **not** as `(Ok(x) | Err(x)): Result<u8, u8>`.  But maybe that difference would be irrelevant because all parts of an or pattern need to have the same type?

mark-i-m commented:
@scottmcm

1. The RFC says that fn args take `pat<no_top_alt>` which requires the parens.

2. I think you are correct. If we want future compatibility with nested type ascription, we may want to change let to also use  `pat<no_top_alt>`. I think none of the others need to change though.

bors commented:
:umbrella: The latest upstream changes (presumably #81761) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

petrochenkov commented:
FWIW, we can always keep the combination "or pattern" + "type ascription" unstable  to unblock or patterns in general and answer the question about priorities later.

bors commented:
:umbrella: The latest upstream changes (presumably #81810) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

Mark-Simulacrum commented:
> It does not stabilize or-patterns in closure arguments or macro :pat matchers (more below).

Am I correct in assuming that this stabilizes or patterns in all places a pattern can go *except* for these two that are specifically called out? I see a listing in the stabilization report, but since there's not language specifically saying "all except for" I'm worried perhaps I overlooked something.

mark-i-m commented:
@Mark-Simulacrum Yes

mark-i-m commented:
Working on https://github.com/rust-lang/rust/pull/82048 to disallow or-patterns before type-ascription, as per @petrochenkov's suggestion.

scottmcm commented:
@rfcbot concern parens-in-let

We [discussed](https://hackmd.io/7f-Y8vrKSkaIecxahDemqQ?view#%E2%80%9CStabilize-or_patterns-RFC-2535-2530-2175%E2%80%9D-rust79278) this in this week's @rust-lang/lang team meeting.  We would like to see all three annotated positions -- function parameters, closure parameters, and lets -- all require the parentheses for now.  (Thus `pat<no_top_alt>` for all of them, like you mentioned in https://github.com/rust-lang/rust/pull/79278#issuecomment-773371179, as I understand it.)

There was not agreement on whether this should be the final state -- specifically some thought that no parens should be needed for the non-annotated case, while others preferred them there too -- but the hope was that using the same grammar production in the three places for now would let the feature stabilize without being too annoying to implement, and then we'll learn how annoying the parens are in practice as more people start using this.

mark-i-m commented:
@scottmcm https://github.com/rust-lang/rust/pull/82048 prohibits top-level or-patterns in `let` bindings, as per you comment. I've also updated the [stabilization report](https://github.com/rust-lang/rust/pull/79278#issuecomment-772815177).

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2034161746) [(plain)](https://github.com/rust-lang/rust/commit/ea78cc1cc772cab54da5adcdbcb7b3a0f17d2a4d/checks/2034161746/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  EXTRA_VARIABLES: {
 "CI_ONLY_WHEN_SUBMODULES_CHANGED": 1
##[endgroup]
adding extra environment variable CI_ONLY_WHEN_SUBMODULES_CHANGED
linux builder detected, using docker to run the build
##[group]Run src/ci/scripts/should-skip-this.sh
---
   Compiling toml v0.5.7
   Compiling url v2.1.1
   Compiling clippy_utils v0.1.52 (/checkout/src/tools/clippy/clippy_utils)
   Compiling cargo_metadata v0.12.0
error: the feature `or_patterns` has been stable since 1.52.0 and no longer requires an attribute to enable
  |
3 | #![feature(or_patterns)]
  |            ^^^^^^^^^^^
  |
  |
  = note: `-D stable-features` implied by `-D warnings`
error: aborting due to previous error

error: could not compile `clippy_utils`

---
   Compiling lsp-types v0.60.0
   Compiling clippy_utils v0.1.52 (/checkout/src/tools/clippy/clippy_utils)
   Compiling structopt v0.3.16
   Compiling cargo_metadata v0.8.2
warning: the feature `or_patterns` has been stable since 1.52.0 and no longer requires an attribute to enable
  |
3 | #![feature(or_patterns)]
  |            ^^^^^^^^^^^
  |
---
   Compiling tokio v0.2.24
   Compiling rustc-rayon v0.3.0
   Compiling tempfile v3.1.0
   Compiling clippy_lints v0.1.52 (/checkout/src/tools/clippy/clippy_lints)
error: top-level or-patterns are not allowed in `let` bindings
    |
    |
888 |         let Sugg::NonParen(s) | Sugg::MaybeParen(s) | Sugg::BinOp(_, s) = &self.0;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: wrap the pattern in parentheses: `(Sugg::NonParen(s) | Sugg::MaybeParen(s) | Sugg::BinOp(_, s))`
   Compiling futures-executor v0.3.12
   Compiling rls-analysis v0.18.1
   Compiling proc-macro-crate v0.1.5
   Compiling tokio-util v0.3.1
---
 Downloading crates ...
  Downloaded getrandom v0.2.0
   Compiling getrandom v0.2.0
   Compiling miri v0.1.0 (/checkout/src/tools/miri)
warning: the feature `or_patterns` has been stable since 1.52.0 and no longer requires an attribute to enable
  |
5 | #![feature(or_patterns)]
  |            ^^^^^^^^^^^
  |
---
   Compiling colored v2.0.0
   Compiling tester v0.7.0
   Compiling rustfix v0.5.1
   Compiling miri v0.1.0 (/checkout/src/tools/miri)
warning: the feature `or_patterns` has been stable since 1.52.0 and no longer requires an attribute to enable
  |
5 | #![feature(or_patterns)]
  |            ^^^^^^^^^^^
  |
  |
  = note: `#[warn(stable_features)]` on by default

warning: the feature `or_patterns` has been stable since 1.52.0 and no longer requires an attribute to enable
  |
5 | #![feature(or_patterns)]
  |            ^^^^^^^^^^^
  |
---
Copying stage2 std from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage2 tool clippy-driver (x86_64-unknown-linux-gnu)
warning: the feature `or_patterns` has been stable since 1.52.0 and no longer requires an attribute to enable
  |
3 | #![feature(or_patterns)]
  |            ^^^^^^^^^^^
  |
  |
  = note: `#[warn(stable_features)]` on by default

warning: 1 warning emitted

   Compiling cargo_metadata v0.12.0
   Compiling clippy_lints v0.1.52 (/checkout/src/tools/clippy/clippy_lints)
error: top-level or-patterns are not allowed in `let` bindings
    |
    |
888 |         let Sugg::NonParen(s) | Sugg::MaybeParen(s) | Sugg::BinOp(_, s) = &self.0;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: wrap the pattern in parentheses: `(Sugg::NonParen(s) | Sugg::MaybeParen(s) | Sugg::BinOp(_, s))`
error: aborting due to previous error

error: could not compile `clippy_lints`


To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/clippy/Cargo.toml" "--features" "rustc-workspace-hack/all-static" "--message-format" "json-render-diagnostics"
expected success, got: exit code: 101
thread 'main' panicked at 'in-tree tool', src/bootstrap/test.rs:533:14
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/clippy
Build completed unsuccessfully in 0:00:12
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2041451063) [(plain)](https://github.com/rust-lang/rust/commit/74d949ca0d635bbfa5b33410a0e639d60d0117a5/checks/2041451063/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking clippy_lints v0.1.52 (/checkout/src/tools/clippy/clippy_lints)
error: unnecessary parentheses around pattern
   --> src/tools/clippy/clippy_lints/src/loops.rs:888:13
    |
888 |         let (Sugg::NonParen(s) | Sugg::MaybeParen(s) | Sugg::BinOp(_, s)) = &self.0;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove these parentheses
    |
    = note: `-D unused-parens` implied by `-D warnings`
error: aborting due to previous error

error: could not compile `clippy_lints`

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #82896) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #82953) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #83074) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

mark-i-m commented:
Should this be renominated? It looks like we still need a few more boxes checked.

joshtriplett commented:
cc @cramertj @nikomatsakis @pnkfelix

nikomatsakis commented:
@rfcbot reviewed

bors commented:
:umbrella: The latest upstream changes (presumably #83301) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

mark-i-m commented:
Rebased.

Also, I changed the stabilization version to 1.53, since the beta will branch before the FCP finishes.

nikomatsakis commented on compiler/rustc_expand/src/expand.rs:
```diff
@@ -917,7 +917,7 @@ pub fn parse_ast_fragment<'a>(
         AstFragmentKind::Ty => AstFragment::Ty(this.parse_ty()?),
         AstFragmentKind::Pat => {
-            AstFragment::Pat(this.parse_pat_allow_top_alt(None, GateOr::Yes, RecoverComma::No)?)
+            AstFragment::Pat(this.parse_pat_allow_top_alt(None, RecoverComma::No)?)
```
I'm a bit confused about this =) does this mean that `pat` accepts `a|b`? I guess I expected some edition logic, and the new fragment matchers, is that in another PR?

mark-i-m replied:
I didn't fully understand this change very well either, but @petrochenkov seems confident (https://github.com/rust-lang/rust/pull/81869#discussion_r573234448). Apparently, this code is for parsing a pattern produced by a macro expansion.

The code you are probably thinking of is here: https://github.com/rust-lang/rust/pull/79278/files#diff-1a054923fe6e121f46954f52f6660a22970818b5c8b3ef1c0f3b703eae2baef6L122-L128

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit b9ecba30bc45af1d5ebe6480ccfb69efc7496a25 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis b9ecba30bc45af1d5ebe6480ccfb69efc7496a25 -->
<!-- homu: {"type":"Approved","sha":"b9ecba30bc45af1d5ebe6480ccfb69efc7496a25","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit b9ecba30bc45af1d5ebe6480ccfb69efc7496a25 with merge 5d04957a4b4714f71d38326fc96a0b0ef6dc5800...
<!-- homu: {"type":"BuildStarted","head_sha":"b9ecba30bc45af1d5ebe6480ccfb69efc7496a25","merge_sha":"5d04957a4b4714f71d38326fc96a0b0ef6dc5800"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2170654267)
Approved by: nikomatsakis
Pushing 5d04957a4b4714f71d38326fc96a0b0ef6dc5800 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2170654267"},"merge_sha":"5d04957a4b4714f71d38326fc96a0b0ef6dc5800"} -->
